inference-test-tool/Dockerfile

```diff
@@ -1,13 +1,32 @@
-FROM python:3.8-slim
+FROM python:3.7-slim
 
 RUN apt-get update
 
+# Install Python3-GDCM (dependency to read pixels of certain DICOM files)
+RUN apt-get install -y python3-gdcm
+RUN cp /usr/lib/python3/dist-packages/gdcm.py /usr/local/lib/python3.7/site-packages/ \
+    && cp /usr/lib/python3/dist-packages/gdcmswig.py /usr/local/lib/python3.7/site-packages/ \
+    && cp /usr/lib/python3/dist-packages/_gdcmswig*.so /usr/local/lib/python3.7/site-packages/ \
+    && cp /usr/lib/x86_64-linux-gnu/libgdcm* /usr/local/lib/python3.7/site-packages/
+
 # Basic env setup
 WORKDIR /opt
 
+# Create User that will run the script. 
+# This is so that the output files are generated with the host user as owner
+ARG USER_ID
+ARG GROUP_ID
+
+RUN if [ ${USER_ID:-0} -ne 0 ] && [ ${GROUP_ID:-0} -ne 0 ]; then \
+    groupadd -g ${GROUP_ID} inference-user &&\
+    useradd -l -u ${USER_ID} -g inference-user inference-user &&\
+    install -d -m 0755 -o inference-user -g inference-user /home/inference-user \
+;fi
+        
+USER inference-user
+
```

Comment:
I couldn't get the docker image to build, unless I moved this part **after** installation steps in `/opt`.

```
ERROR: Could not open requirements file: [Errno 13] Permission denied: '/opt/requirements.txt'
```

But even so, I get an error when running `./send-inference-request.sh`

```
Traceback (most recent call last):
  File "run.py", line 136, in <module>
    upload_study_me(args.file_path, args.segmentation_model, args.host, args.port, args.output)
  File "run.py", line 116, in upload_study_me
    test_inference_boxes.generate_images_with_boxes(images, boxes, output_folder)
  File "/opt/test_inference_boxes.py", line 39, in generate_images_with_boxes
    pil_image.save(output_filename)
  File "/usr/local/lib/python3.7/site-packages/PIL/Image.py", line 2099, in save
    fp = builtins.open(filename, "w+b")
PermissionError: [Errno 13] Permission denied: 'output/0_2.dcm.png'
```

In the end, I had to remove this block to have something building **and** running, with the downside of not owning the output images.

Reply:
Hey @aelkhour I just rebased the branch and tested it again. It works on my Ubuntu. On what OS are you trying it? Also, remember to remove the `test_arterys_sdk` image after changing the branch. Maybe it is not so good that the image is not rebuilt when it exists.

I saw the error you get in `./send-inference-request.sh` once and when checking, the output folder of the test tool was created with root as owner. Not sure how that happened or if that folder was there from before. After removing that folder (test_2d_out in my case) it has worked fine ever since

Reply:
Indeed, the second issue is solved if I remove the output folder beforehand, thanks. 

This said, my remark regarding the build issue still stands, I can't rebuild the docker image unless I make sure all operations in `/opt` are done before changing user.

Reply:
@aelkhour appreciate you trying this out. I also ran into a state issue when attempting to run `python3 -m unittest` on this branch due to I believe having an older image version and it not automatically rebuilding the docker image. 

With this new change @mats-claassen just made  https://github.com/arterys/inference-sdk/pull/8/commits/b88ff835e90c593b9ec696c21e436b2c7931b353 it will always call `docker build` to ensure the latest image is built. This should only slightly increase the invocation time thanks to docker caching and prevent any state issues like this in the future. 

Please let us know if you're still having issues with `python3 -m unittest` after pulling that improvement?

Reply:
@nsvarich I tested the updated PR and I still have the same problem (which occurs at docker build step, so even before being able to run unit tests).

I'm curious, is my original suggestion to move the user definition after python install steps in `/opt` a bad thing? Otherwise I am totally fine with this PR overall.

As this may be related to my local configuration, here are a few more details:
- I am running on Ubuntu 18.04 LTS, with `Docker version 18.09.7, build 2d0083d`.
- Here are my pre-installed images before running the build step.
```
REPOSITORY                                            TAG                 IMAGE ID            CREATED             SIZE
inferencesdk_dev                                      latest              1d36217c6a3a        7 days ago          2.93GB
ubuntu                                                16.04               77be327e4b63        6 weeks ago         124MB
```
- I trigger the test image build by running `./send-inference-request.sh`, and that's where I get the following error:

```
Step 11/13 : RUN python3 -m pip install -r /opt/requirements.txt
 ---> Running in c625ec45619c
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not open requirements file: [Errno 13] Permission denied: '/opt/requirements.txt'
The command '/bin/sh -c python3 -m pip install -r /opt/requirements.txt' returned a non-zero code: 1
```
You can also find below the full build log
```
Sending build context to Docker daemon    488MB
Step 1/13 : FROM python:3.7-slim
3.7-slim: Pulling from library/python
c499e6d256d6: Pull complete 
62b0f1bf7919: Pull complete 
30b7b8e6b50d: Pull complete 
f4947adccc3c: Pull complete 
219c07faa520: Pull complete 
Digest: sha256:e7657d79583a4eaaee8ef642c40099bb6550d447d9e19c08a994dc836eed2327
Status: Downloaded newer image for python:3.7-slim
 ---> 74ac77e9873a
Step 2/13 : RUN apt-get update
 ---> Running in 5a74e404c1e5
Get:1 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:2 http://security.debian.org/debian-security buster/updates/main amd64 Packages [187 kB]
Get:3 http://deb.debian.org/debian buster InRelease [122 kB]
Get:4 http://deb.debian.org/debian buster-updates InRelease [49.3 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 Packages [7907 kB]
Get:6 http://deb.debian.org/debian buster-updates/main amd64 Packages [7380 B]
Fetched 8337 kB in 6s (1423 kB/s)
Reading package lists...
Removing intermediate container 5a74e404c1e5
 ---> ce6125518cb0
Step 3/13 : RUN apt-get install -y python3-gdcm
 ---> Running in 76de92ea2567
Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  bzip2 file libcharls2 libgdcm2.8 libjson-c3 libmagic-mgc libmagic1 libmpdec2
  libopenjp2-7 libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib
  libsocket++1 mime-support python3 python3-minimal python3.7
  python3.7-minimal xz-utils
Suggested packages:
  bzip2-doc python3-doc python3-tk python3-venv python3.7-venv python3.7-doc
  binutils binfmt-support
The following NEW packages will be installed:
  bzip2 file libcharls2 libgdcm2.8 libjson-c3 libmagic-mgc libmagic1 libmpdec2
  libopenjp2-7 libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib
  libsocket++1 mime-support python3 python3-gdcm python3-minimal python3.7
  python3.7-minimal xz-utils
0 upgraded, 20 newly installed, 0 to remove and 1 not upgraded.
Need to get 8045 kB of archives.
After this operation, 45.9 MB of additional disk space will be used.
Get:1 http://deb.debian.org/debian buster/main amd64 libpython3.7-minimal amd64 3.7.3-2+deb10u1 [589 kB]
Get:2 http://deb.debian.org/debian buster/main amd64 python3.7-minimal amd64 3.7.3-2+deb10u1 [1736 kB]
Get:3 http://deb.debian.org/debian buster/main amd64 python3-minimal amd64 3.7.3-1 [36.6 kB]
Get:4 http://deb.debian.org/debian buster/main amd64 mime-support all 3.62 [37.2 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 libmpdec2 amd64 2.4.2-2 [87.2 kB]
Get:6 http://deb.debian.org/debian buster/main amd64 libpython3.7-stdlib amd64 3.7.3-2+deb10u1 [1734 kB]
Get:7 http://deb.debian.org/debian buster/main amd64 python3.7 amd64 3.7.3-2+deb10u1 [330 kB]
Get:8 http://deb.debian.org/debian buster/main amd64 libpython3-stdlib amd64 3.7.3-1 [20.0 kB]
Get:9 http://deb.debian.org/debian buster/main amd64 python3 amd64 3.7.3-1 [61.5 kB]
Get:10 http://deb.debian.org/debian buster/main amd64 bzip2 amd64 1.0.6-9.2~deb10u1 [48.4 kB]
Get:11 http://deb.debian.org/debian buster/main amd64 libmagic-mgc amd64 1:5.35-4+deb10u1 [242 kB]
Get:12 http://deb.debian.org/debian buster/main amd64 libmagic1 amd64 1:5.35-4+deb10u1 [117 kB]
Get:13 http://deb.debian.org/debian buster/main amd64 file amd64 1:5.35-4+deb10u1 [66.4 kB]
Get:14 http://deb.debian.org/debian buster/main amd64 xz-utils amd64 5.2.4-1 [183 kB]
Get:15 http://deb.debian.org/debian buster/main amd64 libcharls2 amd64 2.0.0+dfsg-1 [64.3 kB]
Get:16 http://deb.debian.org/debian buster/main amd64 libjson-c3 amd64 0.12.1+ds-2 [26.9 kB]
Get:17 http://deb.debian.org/debian buster/main amd64 libopenjp2-7 amd64 2.3.0-2+deb10u1 [158 kB]
Get:18 http://deb.debian.org/debian buster/main amd64 libsocket++1 amd64 1.12.13-10 [42.1 kB]
Get:19 http://deb.debian.org/debian buster/main amd64 libgdcm2.8 amd64 2.8.8-9 [1834 kB]
Get:20 http://deb.debian.org/debian buster/main amd64 python3-gdcm amd64 2.8.8-9 [630 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 8045 kB in 1s (9510 kB/s)
Selecting previously unselected package libpython3.7-minimal:amd64.
(Reading database ... 6828 files and directories currently installed.)
Preparing to unpack .../libpython3.7-minimal_3.7.3-2+deb10u1_amd64.deb ...
Unpacking libpython3.7-minimal:amd64 (3.7.3-2+deb10u1) ...
Selecting previously unselected package python3.7-minimal.
Preparing to unpack .../python3.7-minimal_3.7.3-2+deb10u1_amd64.deb ...
Unpacking python3.7-minimal (3.7.3-2+deb10u1) ...
Setting up libpython3.7-minimal:amd64 (3.7.3-2+deb10u1) ...
Setting up python3.7-minimal (3.7.3-2+deb10u1) ...
Selecting previously unselected package python3-minimal.
(Reading database ... 7069 files and directories currently installed.)
Preparing to unpack .../0-python3-minimal_3.7.3-1_amd64.deb ...
Unpacking python3-minimal (3.7.3-1) ...
Selecting previously unselected package mime-support.
Preparing to unpack .../1-mime-support_3.62_all.deb ...
Unpacking mime-support (3.62) ...
Selecting previously unselected package libmpdec2:amd64.
Preparing to unpack .../2-libmpdec2_2.4.2-2_amd64.deb ...
Unpacking libmpdec2:amd64 (2.4.2-2) ...
Selecting previously unselected package libpython3.7-stdlib:amd64.
Preparing to unpack .../3-libpython3.7-stdlib_3.7.3-2+deb10u1_amd64.deb ...
Unpacking libpython3.7-stdlib:amd64 (3.7.3-2+deb10u1) ...
Selecting previously unselected package python3.7.
Preparing to unpack .../4-python3.7_3.7.3-2+deb10u1_amd64.deb ...
Unpacking python3.7 (3.7.3-2+deb10u1) ...
Selecting previously unselected package libpython3-stdlib:amd64.
Preparing to unpack .../5-libpython3-stdlib_3.7.3-1_amd64.deb ...
Unpacking libpython3-stdlib:amd64 (3.7.3-1) ...
Setting up python3-minimal (3.7.3-1) ...
Selecting previously unselected package python3.
(Reading database ... 7506 files and directories currently installed.)
Preparing to unpack .../00-python3_3.7.3-1_amd64.deb ...
Unpacking python3 (3.7.3-1) ...
Selecting previously unselected package bzip2.
Preparing to unpack .../01-bzip2_1.0.6-9.2~deb10u1_amd64.deb ...
Unpacking bzip2 (1.0.6-9.2~deb10u1) ...
Selecting previously unselected package libmagic-mgc.
Preparing to unpack .../02-libmagic-mgc_1%3a5.35-4+deb10u1_amd64.deb ...
Unpacking libmagic-mgc (1:5.35-4+deb10u1) ...
Selecting previously unselected package libmagic1:amd64.
Preparing to unpack .../03-libmagic1_1%3a5.35-4+deb10u1_amd64.deb ...
Unpacking libmagic1:amd64 (1:5.35-4+deb10u1) ...
Selecting previously unselected package file.
Preparing to unpack .../04-file_1%3a5.35-4+deb10u1_amd64.deb ...
Unpacking file (1:5.35-4+deb10u1) ...
Selecting previously unselected package xz-utils.
Preparing to unpack .../05-xz-utils_5.2.4-1_amd64.deb ...
Unpacking xz-utils (5.2.4-1) ...
Selecting previously unselected package libcharls2:amd64.
Preparing to unpack .../06-libcharls2_2.0.0+dfsg-1_amd64.deb ...
Unpacking libcharls2:amd64 (2.0.0+dfsg-1) ...
Selecting previously unselected package libjson-c3:amd64.
Preparing to unpack .../07-libjson-c3_0.12.1+ds-2_amd64.deb ...
Unpacking libjson-c3:amd64 (0.12.1+ds-2) ...
Selecting previously unselected package libopenjp2-7:amd64.
Preparing to unpack .../08-libopenjp2-7_2.3.0-2+deb10u1_amd64.deb ...
Unpacking libopenjp2-7:amd64 (2.3.0-2+deb10u1) ...
Selecting previously unselected package libsocket++1:amd64.
Preparing to unpack .../09-libsocket++1_1.12.13-10_amd64.deb ...
Unpacking libsocket++1:amd64 (1.12.13-10) ...
Selecting previously unselected package libgdcm2.8:amd64.
Preparing to unpack .../10-libgdcm2.8_2.8.8-9_amd64.deb ...
Unpacking libgdcm2.8:amd64 (2.8.8-9) ...
Selecting previously unselected package python3-gdcm.
Preparing to unpack .../11-python3-gdcm_2.8.8-9_amd64.deb ...
Unpacking python3-gdcm (2.8.8-9) ...
Setting up libjson-c3:amd64 (0.12.1+ds-2) ...
Setting up mime-support (3.62) ...
Setting up libmagic-mgc (1:5.35-4+deb10u1) ...
Setting up libcharls2:amd64 (2.0.0+dfsg-1) ...
Setting up libmagic1:amd64 (1:5.35-4+deb10u1) ...
Setting up file (1:5.35-4+deb10u1) ...
Setting up bzip2 (1.0.6-9.2~deb10u1) ...
Setting up xz-utils (5.2.4-1) ...
update-alternatives: using /usr/bin/xz to provide /usr/bin/lzma (lzma) in auto mode
update-alternatives: warning: skip creation of /usr/share/man/man1/lzma.1.gz because associated file /usr/share/man/man1/xz.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/unlzma.1.gz because associated file /usr/share/man/man1/unxz.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzcat.1.gz because associated file /usr/share/man/man1/xzcat.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzmore.1.gz because associated file /usr/share/man/man1/xzmore.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzless.1.gz because associated file /usr/share/man/man1/xzless.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzdiff.1.gz because associated file /usr/share/man/man1/xzdiff.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzcmp.1.gz because associated file /usr/share/man/man1/xzcmp.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzgrep.1.gz because associated file /usr/share/man/man1/xzgrep.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzegrep.1.gz because associated file /usr/share/man/man1/xzegrep.1.gz (of link group lzma) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/lzfgrep.1.gz because associated file /usr/share/man/man1/xzfgrep.1.gz (of link group lzma) doesn't exist
Setting up libsocket++1:amd64 (1.12.13-10) ...
Setting up libopenjp2-7:amd64 (2.3.0-2+deb10u1) ...
Setting up libmpdec2:amd64 (2.4.2-2) ...
Setting up libpython3.7-stdlib:amd64 (3.7.3-2+deb10u1) ...
Setting up libgdcm2.8:amd64 (2.8.8-9) ...
Setting up libpython3-stdlib:amd64 (3.7.3-1) ...
Setting up python3.7 (3.7.3-2+deb10u1) ...
Setting up python3 (3.7.3-1) ...
running python rtupdate hooks for python3.7...
running python post-rtupdate hooks for python3.7...
Setting up python3-gdcm (2.8.8-9) ...
Processing triggers for libc-bin (2.28-10) ...
Removing intermediate container 76de92ea2567
 ---> 3ffa85a06db9
Step 4/13 : RUN cp /usr/lib/python3/dist-packages/gdcm.py /usr/local/lib/python3.7/site-packages/     && cp /usr/lib/python3/dist-packages/gdcmswig.py /usr/local/lib/python3.7/site-packages/     && cp /usr/lib/python3/dist-packages/_gdcmswig*.so /usr/local/lib/python3.7/site-packages/     && cp /usr/lib/x86_64-linux-gnu/libgdcm* /usr/local/lib/python3.7/site-packages/
 ---> Running in 0e2df05c8795
Removing intermediate container 0e2df05c8795
 ---> ef91a7968d5d
Step 5/13 : WORKDIR /opt
 ---> Running in 80c1b190b1ad
Removing intermediate container 80c1b190b1ad
 ---> 4e7a7549079e
Step 6/13 : ARG USER_ID
 ---> Running in 91880f396e56
Removing intermediate container 91880f396e56
 ---> 7c36dbfd5332
Step 7/13 : ARG GROUP_ID
 ---> Running in 6513865e50f7
Removing intermediate container 6513865e50f7
 ---> 8f80effb703e
Step 8/13 : RUN if [ ${USER_ID:-0} -ne 0 ] && [ ${GROUP_ID:-0} -ne 0 ]; then     groupadd -g ${GROUP_ID} inference-user &&    useradd -l -u ${USER_ID} -g inference-user inference-user &&    install -d -m 0755 -o inference-user -g inference-user /home/inference-user ;fi
 ---> Running in f04ace782fa5
Removing intermediate container f04ace782fa5
 ---> 3805e7b29246
Step 9/13 : USER inference-user
 ---> Running in 99e64f223f8f
Removing intermediate container 99e64f223f8f
 ---> 0232f88e5789
Step 10/13 : COPY requirements.txt /opt/requirements.txt
 ---> 0e281e17677b
Step 11/13 : RUN python3 -m pip install -r /opt/requirements.txt
 ---> Running in c625ec45619c
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not open requirements file: [Errno 13] Permission denied: '/opt/requirements.txt'
The command '/bin/sh -c python3 -m pip install -r /opt/requirements.txt' returned a non-zero code: 1
Unable to find image 'test_arterys_sdk:latest' locally
docker: Error response from daemon: pull access denied for test_arterys_sdk, repository does not exist or may require 'docker login'.
See 'docker run --help'.
```

Reply:
@aelkhour looking at your log dump above, you may have changed the base image from `FROM ubuntu:16.04` to `FROM python:3.7-slim`?

I think we had made that switch for the inference-test-tools container, but we kept the server container an ubuntu one.

Reply:
@aelkhour I can't seem to reproduce. Perhaps try removing the `test_arterys_sdk` image and then running `docker image build --build-arg USER_ID=$(id -u ${USER}) --build-arg GROUP_ID=$(id -g ${USER}) -t test_arterys_sdk .` again if you haven't already tried?

Reply:
@bsblaw I tried and I still get the same error at step 11/13...

I don't want to be a burden any further, if many of you have tested the MR and seen it work, there's something I'm doing wrong on my end and you should probably merge.

Reply:
@aelkhour Your feedback is valuable. The only meaningful difference I see from your setup is the Docker version. Is there a reason why you are not using a newer one like Docker `19.03`?

Reply:
I am using the default version of Docker as provided in Ubuntu 18.04, installed with `sudo apt-get install docker`, so I did not see the point of upgrading.